src/ccs/multifolding.rs

```diff
@@ -0,0 +1,390 @@
+
+    // (sum gamma^j * e1 * sigma_j)
+    for (j, sigma_j) in sigmas.iter().enumerate() {
+      let gamma_j = gamma.pow([j as u64]);
+      c += gamma_j * e1 * sigma_j;
+    }
+
+    // + gamma^{t+1} * e2 * sum c_i * prod theta_j
+    let mut lhs = G::Scalar::ZERO;
+    for i in 0..self.ccs.q {
+      let mut prod = G::Scalar::ONE;
+      for j in self.ccs.S[i].clone() {
+        prod *= thetas[j];
+      }
+      lhs += self.ccs.c[i] * prod;
+    }
+    let gamma_t1 = gamma.pow([(self.ccs.t + 1) as u64]);
+    c += gamma_t1 * e2 * lhs;
+    c
+  }
+
+  /// Compute g(x) polynomial for the given inputs.
+  pub fn compute_g(
+    running_instance: &LCCCS<G>,
+    cccs_instance: &CCCSShape<G>,
+    z1: &Vec<G::Scalar>,
+    z2: &Vec<G::Scalar>,
+    gamma: G::Scalar,
+    beta: &[G::Scalar],
+  ) -> VirtualPolynomial<G::Scalar> {
+    let mut vec_L = running_instance.compute_Ls(z1);
+    let mut Q = cccs_instance
+      .compute_Q(z2, beta)
+      .expect("TQ comp should not fail");
+    let mut g = vec_L[0].clone();
+    for (j, L_j) in vec_L.iter_mut().enumerate().skip(1) {
+      let gamma_j = gamma.pow([j as u64]);
+      L_j.scalar_mul(&gamma_j);
+      g = g.add(L_j);
+    }
+    let gamma_t1 = gamma.pow([(cccs_instance.ccs.t + 1) as u64]);
+    Q.scalar_mul(&gamma_t1);
+    g = g.add(&Q);
+    g
+  }
+
+  // XXX: This might need to be mutable if we want to hold an LCCCS instance as the IVC inside the
+  // NIMFS object.
+  pub fn fold(
+    &self,
+    lcccs1: &LCCCS<G>,
+    cccs2: &CCCSInstance<G>,
+    sigmas: &[G::Scalar],
+    thetas: &[G::Scalar],
+    r_x_prime: Vec<G::Scalar>,
+    rho: G::Scalar,
+  ) -> LCCCS<G> {
+    let C = lcccs1.C + cccs2.C.mul(rho);
+    let u = lcccs1.u + rho;
+    let x: Vec<G::Scalar> = lcccs1
+      .x
+      .iter()
+      .zip(
+        cccs2
+          .x
+          .iter()
+          .map(|x_i| *x_i * rho)
+          .collect::<Vec<G::Scalar>>(),
+      )
+      .map(|(a_i, b_i)| *a_i + b_i)
+      .collect();
+    let v: Vec<G::Scalar> = sigmas
+      .iter()
+      .zip(
+        thetas
+          .iter()
+          .map(|x_i| *x_i * rho)
+          .collect::<Vec<G::Scalar>>(),
+      )
+      .map(|(a_i, b_i)| *a_i + b_i)
+      .collect();
+
+    LCCCS {
+      matrix_mles: lcccs1.matrix_mles.clone(),
+      C,
+      ccs: lcccs1.ccs.clone(),
+      u,
+      x,
+      r_x: r_x_prime,
+      v,
+    }
+  }
+
+  pub fn fold_witness(w1: &CCSWitness<G>, w2: &CCSWitness<G>, rho: G::Scalar) -> CCSWitness<G> {
+    let w = w1
+      .w
+      .iter()
+      .zip(
+        w2.w
+          .iter()
+          .map(|x_i| *x_i * rho)
+          .collect::<Vec<G::Scalar>>(),
+      )
+      .map(|(a_i, b_i)| *a_i + b_i)
+      .collect();
+
+    // XXX: There's no handling of r_w atm. So we will ingore until all folding is implemented,
+    // let r_w = w1.r_w + rho * w2.r_w;
+    CCSWitness { w }
+  }
+}
+
+/// Evaluate eq polynomial.
+pub fn eq_eval<F: PrimeField>(x: &[F], y: &[F]) -> F {
```

Comment:
There is a type called EqPolynomial in src/spartan/polynomial.rs. Should we use that?